src/cargo/sources/git/utils.rs

```diff
@@ -358,66 +359,169 @@ impl<'a> GitCheckout<'a> {
+    let mut username_attempts = Vec::new();
+
     let res = f(&mut |url, username, allowed| {
-        called += 1;
-        if called >= 2 {
-            return Err(git2::Error::from_str("no authentication available"))
+        let allowed = allowed & !attempted;
+
+        // libgit2's "USERNAME" authentication actually means that it's just
+        // asking us for a username to keep going. This is currently only really
+        // used for SSH authentication and isn't really an authentication type.
+        // The logic currently looks like:
+        //
+        //      let user = ...;
+        //      if (user.is_null())
+        //          user = callback(USERNAME, null, ...);
+        //
+        //      callback(SSH_KEY, user, ...)
+        //
+        // So if we have a USERNAME request we just pass it either `username` or
+        // a fallback of "git". We'll do some more principled attempts later on.
+        if allowed.contains(git2::USERNAME) {
+            attempted = attempted | git2::USERNAME;
+            return git2::Cred::username(username.unwrap_or("git"))
         }
-        if allowed.contains(git2::SSH_KEY) ||
-                       allowed.contains(git2::USERNAME) {
-            let user = username.map(|s| s.to_string())
-                               .or_else(|| cred_helper.username.clone())
-                               .unwrap_or("git".to_string());
-            if allowed.contains(git2::USERNAME) {
-                git2::Cred::username(&user)
-            } else {
-                git2::Cred::ssh_key_from_agent(&user)
+
+        // An "SSH_KEY" authentication indicates that we need some sort of SSH
+        // authentication. This can currently either come from the ssh-agent
+        // process or from a raw in-memory SSH key. Cargo only supports using
+        // ssh-agent currently.
+        //
+        // We try a few different usernames here, including:
+        //
+        //  1. The `username` argument, if provided. This will cover cases where
+        //     the user was passed in the URL, for example.
+        //  2. The global credential helper's username, if any is configured
+        //  3. The local account's username (if present)
+        //  4. Finally, "git" as it's a common fallback (e.g. with github)
+        if allowed.contains(git2::SSH_KEY) {
+            loop {
+                let name = match username_attempt {
+                    UsernameAttempt::Arg => {
+                        username_attempt = UsernameAttempt::CredHelper;
+                        username.map(|s| s.to_string())
+                    }
+                    UsernameAttempt::CredHelper => {
+                        username_attempt = UsernameAttempt::Local;
+                        cred_helper.username.clone()
+                    }
+                    UsernameAttempt::Local => {
+                        username_attempt = UsernameAttempt::Git;
+                        env::var("USER").ok()
```

Comment:
On Windows, the local username is stored in the `%USERNAME%` environment variable.  `%USER%` is not generally set.